### DISCUSSION TITLE
Asynchronous calls in validation

### DISCUSSION THREAD

**Discussion by HonzaMac** [markdoc/markdoc#176] [Q&A]
Thank you for recent release of async [experimental feature in 0.1.4](https://github.com/markdoc/markdoc/releases/tag/0.1.4).

We have tried it and it works. We could rapidly simplify our code for uploading images.

I think that `validate(node, config)` should be also `async`. 
For use-case of image upload, we can validate existence of file in validation phase. User would then get validation error on non-existent file instead of waiting for transformation and stop uploading of many files in the middle.

What we currently **can't do**: 
```typescript
import {stat} from 'fs/promises';

export const replaceImagesOnTransform = (accessToken: string, directory: string): Schema => ({
  render: TAG_NAME,
  attributes: {
    src: {type: String, required: true},
    alt: {type: String},
  },
  async validate(node, config) { // THIS IS NEW, async
    const {src} = node.transformAttributes(config);
    try {
      await stat(src, {throwIfNoEntry: true});
    } catch (e) {
      return [{id: 'file-not-found', message: `File "${src}" not found. ${e?.message}`, level: 'error'}];
    }
  },
  async transform(node, config) {
    const attributes = node.transformAttributes(config);
    const location = getLocation(attributes.src, directory);
    const src = await uploadSingleFile(location, accessToken);

    return new Tag(TAG_NAME, {...attributes, src}, node.transformChildren(config));
  },
});
```

Any thoughts?

**Comment by mfix-stripe** [comment]
For what it's worth, you can use `fs.statSync` and avoid async altogether. I understand that isn't a viable solution for everything. The team will definitely consider this further.

**Reply by HonzaMac** [reply]
I would rather not do this as it blocks event-loop.

Thank you for Pull request with changes to async, I love it!

**Reply by matv-stripe** [reply]
@HonzaMac is there a way you could gather a list of all possible images and then pass through through the `config`? This is how we handle image validation on stripe.com/docs

**Reply by HonzaMac** [reply]
@matv-stripe Yes, we have it done that way in first place before async was enabled. We gathered image locations in validation, then uploaded and then replaced locations in transform. 

I feel that async validation is better as we can tell user right away that there is something wrong with his data. Eventually once there is a support for Markdoc in IDE, user would see that path to image is invalid before processing and uploading to server(our case).
